### PR TITLE
Update ode.py

### PR DESCRIPTION
**Pull request by Teut2711**
#### Release Notes

* solvers
    * ode
        * added a wrapper
* functions
    *  2 functions added

### CODE REVIEW

**Comment by XtremeGood** [comment]
Hi, Hope you can see the changes now.

**Comment by oscarbenjamin** [comment]
Hi @XtremeGood we don't want to use a decorator for this. This is one of many possible methods used in solving ODEs and decorators don't scale very well if we need many of them. Please carefully read what I said here: [https://github.com/sympy/sympy/pull/15922#issuecomment-460345893](https://github.com/sympy/sympy/pull/15922#issuecomment-460345893)

**Comment by oscarbenjamin** [comment]
You are trying to get your code integrated into the ode module which is almost 10000 lines of code already. Your new code needs to fit in with the rest of that so it needs to be written a particular way. If you take time just to read through the code in the ode module you might learn more about how it fits together.

If you try it and get stuck then you can push the code here and I'll take a look. (I have pretty clear idea about how this can be done.)

Can you please comment using the web interface rather than by email? It's okay if you can't reply immediately.

**Comment by XtremeGood** [comment]
Hi,
Yep,that's why I will try today again.For the time being I 'm preparing for a test.We've got surprised tests.It didn't happen yesterday, it may or may not happen today.

**Comment by XtremeGood** [comment]
Hi,
Sorry I had to study for the test but now I 'm free for the weekend and can get back to work now.

**Comment by XtremeGood** [comment]
Hi,
Sorry for keeping you wait so long.Can you please check the new commit?

Sent from Mail for Windows 10

**Comment by oscarbenjamin** [comment]
This looks better.

Now that you have updated the PR the tests are running. The Travis tests are not completed yet but if you look now you can already see that some are failing. Click the "Details" button below:.

I looked quickly and the errors I saw are because the function `order_reducing_substitution` should have an underscore at the start of its name. You're calling it as `_ order_reducing_substitution`.

**Comment by XtremeGood** [comment]
Thanks for informing. I've changed that.

File: sympy/solvers/ode.py [hunk 1]
```diff
@@ -4003,6 +4002,31 @@ def _frobenius(n, m, p0, q0, p, q, x0, x, c, check=None):
 
     return frobdict
 
+
+def _check_substitution_type(eq, func):
+    # Check if some derivate of positive order can be substituted as g(x) = f^(order_to_subs) (x)
+    x = func.args[0]
+    f = func.func
+    order_to_subs = 0
+    for nsubs_try in range(1,ode_order(eq)):
```
Comment (smichr) [inline sympy/solvers/ode.py]: Suggest `D = Dummy()` before loop and then reuse in loop.

**Comment by oscarbenjamin** [comment]
There are still more errors. You can see them if you click the "Details" button below. I think you would notice the errors before committing if you tested the code yourself though.

Try the example from #15881. If that isn't working then the code isn't ready.

File: sympy/solvers/ode.py (hunk 1 above)
Comment (XtremeGood) [reply sympy/solvers/ode.py]: Did it.

**Comment by XtremeGood** [comment]
@oscarbenjamin 
Yep, I have heard about performing the unit tests before submitting any code but this seemed to be a simple code and I've just switched from default python idle to visual studio code.I will soon learn all that stuff.

**Comment by oscarbenjamin** [comment]
I'm not just talking about unit tests. The errors I can see in the Travis logs are from mistakes that you would have seen if you had tested this with a single example.

If I test your code to try the ODE from #15881 I get this:
```julia
In [1]: eq = f(x).diff(x, 2) + x * f(x).diff(x)**2                                                                                

In [2]: dsolve(eq)                                                                                                                
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-42dc4d2e7b25> in <module>
----> 1 dsolve(eq)

~/current/sympy/sympy/sympy/solvers/ode.py in dsolve(eq, func, hint, simplify, ics, xi, eta, x0, n, **kwargs)
    645         hints = _desolve(eq, func=func,
    646             hint=hint, simplify=True, xi=xi, eta=eta, type='ode', ics=ics,
--> 647             x0=x0, n=n, **kwargs)
    648 
    649         eq = hints.pop('eq', eq)

~/current/sympy/sympy/sympy/solvers/deutils.py in _desolve(eq, func, hint, ics, simplify, **kwargs)
    203     if kwargs.get('classify', True):
    204         hints = classifier(eq, func, dict=True, ics=ics, xi=xi, eta=eta,
--> 205         n=terms, x0=x0, prep=prep)
    206 
    207     else:

~/current/sympy/sympy/sympy/solvers/ode.py in classify_ode(eq, func, dict, ics, **kwargs)
   1360         # integrals e.g.:
   1361         # d^3/dx^3(x y) = F(x)
-> 1362         r = _check_substitution_type(reduced_eq, func)
   1363         if r["solutions"]:
   1364             matching_hints['order_reducing_substitution'] = r

~/current/sympy/sympy/sympy/solvers/ode.py in _check_substitution_type(eq, func)
   4010     order_to_subs = 0
   4011     D = Dummy()
-> 4012     for nsubs_try in range(1,ode_order(eq)):
   4013         if reduced_eq.subs(f(x).diff(x, nsubs_try), D).has(f(x)):
   4014             break

TypeError: ode_order() missing 1 required positional argument: 'func'
```

**Comment by XtremeGood** [comment]
It will take time.It needs a lot of modifications.

**Comment by oscarbenjamin** [comment]
Okay well take your time. Did you try the example I just showed before sending the last two commits?

Test the code locally (on your computer). Don't push it here until you think it might be working.

**Comment by XtremeGood** [comment]
Code breaks at _desolve in utils but that it connected to classify_ode.I got 1 bug though.

**Comment by oscarbenjamin** [comment]
If you get stuck you can push what you have so far and ask for help.

**Comment by XtremeGood** [comment]
I tried it yesterday again but messed the repo. I have decided to give it
limited time otherwise I wouldn't be able to study my subjects in college
and other things which I want to learn like ml, web scrap etc and want to
clear gsoc so that I can make way to the coding world since I have chemical
engineering course here.

**Comment by XtremeGood** [comment]
Hi,
I ran the code in debug mode in visual studio debug mode.It's skipping my function and not including it in hints.I ran debug on dsolve, classify_ode, and _desolve. Since the debug mode works only on one file I couldn't see how dsolve calls _dsolve which futher calls classify_ode. I' m stuck.

**Comment by oscarbenjamin** [comment]
If you push what you have right now I can probably tell you why it's not working.

Can you also show what you tried and what the error message was (the whole error message with traceback)? The traceback should show you which line of which function is calling which other function.

Also you can search for function names within each .py file to see where they get called.

**Comment by XtremeGood** [comment]
Error in : 
hints = _desolve(eq, func=func,
            hint=hint, simplify=True, xi=xi, eta=eta, type='ode', ics=ics,
            x0=x0, n=n, **kwargs)
Error:Exception has occurred: NotImplementedError
solve: Cannot solve x*Derivative(f(x), x)**2 + Derivative(f(x), x, x)
  File "C:\Users\Dell\Desktop\GitHub\sympy\sympy\solvers\ode.py", line 647, in dsolve
    x0=x0, n=n, **kwargs)
  File "C:\Users\Dell\Desktop\GitHub\sympy\sympy\solvers\ode.py", line 8709, in <module>
    print(dsolve(f.diff(x,2)+x*f.diff(x)**2))

**Comment by XtremeGood** [comment]
I tried running the function _desolve. There it directly just to "if not hints['default']".

**Comment by XtremeGood** [comment]
This ->>>Also you can search for function names within each .py file to see where they get called.


Search where?

**Comment by oscarbenjamin** [comment]
When pasting something like an error message or some code you should use backticks for markup: [https://guides.github.com/features/mastering-markdown/](https://guides.github.com/features/mastering-markdown/)

When I said you can search the function name I just meant in your editor. When looking at a file you can search within that file (presumably - I don't know what editor you're using).

File: sympy/solvers/ode.py [hunk 2]
```diff
@@ -4003,6 +4002,33 @@ def _frobenius(n, m, p0, q0, p, q, x0, x, c, check=None):
...
+    for nsubs_try in range(1,ode_order(eq,f(x))):
+        if eq.subs(f(x).diff(x, nsubs_try), D).has(f(x)):
+            break
+        else:
+            order_to_subs += 1
+    if order_to_subs > 0:
+        return {'var':order_to_subs, 'solutions':True}
+
+# Use repeated substitution until we do not have a function independent of derivative   
+def order_reducing_substitution(eq, func, order, match):
```
Comment (oscarbenjamin) [inline sympy/solvers/ode.py]: This function should be called `ode_order_reducing_substitution`

**Comment by XtremeGood** [comment]
I'm using visual studio code.I have heard about logging, and storing error logs but have never used  them. I will have to learn it.

File: sympy/solvers/ode.py (hunk 2 above)
Comment (XtremeGood) [reply sympy/solvers/ode.py]: Why?I mean all other functions are named normally and everywhere I have used this name only.

Comment (XtremeGood) [reply sympy/solvers/ode.py]: And instead of markdown syntax does dpaste.com works?

Comment (oscarbenjamin) [reply sympy/solvers/ode.py]: If the hint is called X then the corresponding function should be called ode_X.

Comment (oscarbenjamin) [reply sympy/solvers/ode.py]: If the hint is called X then the corresponding function should be called ode_X.

Comment (oscarbenjamin) [reply sympy/solvers/ode.py]: No idea about dpaste.com. You should learn the markdown syntax to use github though

Comment (XtremeGood) [reply sympy/solvers/ode.py]: # Use repeated substitution until we do not have a function independent of derivative   
* def ode_order_reducing_substitution(eq, func, order, match):

Comment (XtremeGood) [reply sympy/solvers/ode.py]: ""Use repeated substitution until we do not have a function independent of derivative   ""
'''def ode_order_reducing_substitution(eq, func, order, match):'''

Comment (XtremeGood) [reply sympy/solvers/ode.py]: # Use repeated substitution until we do not have a function independent of derivative   
def ode_order_reducing_substitution(eq, func, order, match):

Comment (XtremeGood) [reply sympy/solvers/ode.py]: -[x] " Use repeated substitution until we do not have a function independent of derivative  " 
-[x]  def ode_order_reducing_substitution(eq, func, order, match):

Comment (XtremeGood) [reply sympy/solvers/ode.py]: Sorry, it will take multiple efforts to learn that styling. I will read that doc tomorrow.

Comment (XtremeGood) [reply sympy/solvers/ode.py]: One thing is that the r""" ....""" is coloring whole of the whole below it orange in editor which happened for quoted text in my editor.I removed the r and the code turned from orange to multi-colored.Strange isn't it?It seems as if the quotes aren't being actually closed even after closing them with """.On the other hand while debugging I could see the function names in the debugging window.I don't know if they were from the all_hints list or somewhere else but this is somewhat strange.

**Comment by oscarbenjamin** [comment]
You don't need to post things to see what they will look like. As I write this just above there is a "preview" tab. If I click that then it shows what the markdown I am editing will look like after posting.

The issue with colouring in your editor is probably specific to whichever editor you are using.

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,70 @@
+{
```
Comment (asmeurer) [inline .vscode/launch.json]: This file should not be here.

File: sympy/solvers/ode.py [hunk 4]
```diff
@@ -8676,3 +8702,8 @@ def _nonlinear_3eq_order1_type5(x, y, t, eq):
     sol2 = dsolve(diff(v(t),t) - (v*(a*F3-c*F1)).subs(u,x_y).subs(w,z_y).subs(v,v(t))).rhs
     sol3 = dsolve(diff(w(t),t) - (w*(b*F1-a*F2)).subs(u,x_z).subs(v,y_z).subs(w,w(t))).rhs
     return [sol1, sol2, sol3]
+
+
+x = Symbol("x")    
+f = Function("f")(x)
+print(dsolve(f.diff(x,2)+x*f.diff(x)**2))
```
Comment (asmeurer) [inline sympy/solvers/ode.py]: This code should not be here.

**Comment by asmeurer** [comment]
The release notes need to be fixed.

File: .vscode/launch.json (hunk 3 above)
Comment (XtremeGood) [reply .vscode/launch.json]: Hi
I 'll delete them.

**Pull request closed without merging**